src/network/networkd-sysctl.c

```diff
@@ -28,22 +30,81 @@ static void manager_set_ip_forwarding(Manager *manager, int family) {
                 return; /* keep */
 
         /* First, set the default value. */
-        r = sysctl_write_ip_property_boolean(family, "default", "forwarding", t);
+        r = sysctl_write_ip_property_boolean(family, "default", "forwarding", t, &manager->sysctl_shadow);
         if (r < 0)
                 log_warning_errno(r, "Failed to %s the default %s forwarding: %m",
                                   enable_disable(t), af_to_ipv4_ipv6(family));
 
         /* Then, set the value to all interfaces. */
-        r = sysctl_write_ip_property_boolean(family, "all", "forwarding", t);
+        r = sysctl_write_ip_property_boolean(family, "all", "forwarding", t, &manager->sysctl_shadow);
         if (r < 0)
                 log_warning_errno(r, "Failed to %s %s forwarding for all interfaces: %m",
                                   enable_disable(t), af_to_ipv4_ipv6(family));
 }
 
+static int sysctl_inotify(sd_event_source *source, const struct inotify_event *event, void *userdata) {
+        _cleanup_free_ char *buf = NULL, *path = NULL;
+        Hashmap *sysctl_shadow = ASSERT_PTR(userdata);
+        const char *dir_path, *value;
+        int r;
+
+        assert(event);
+
+        r = sd_event_source_get_inotify_path(source, &dir_path);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get the watched inotify path: %m");
+
+        path = path_join(dir_path, event->name);
+        if (!path)
+                return log_oom_debug();
+
+        value = hashmap_get(sysctl_shadow, path);
+
+        /* We don't handle this sysctl */
+        if (!value)
+                return r;
+
+        r = read_virtual_file(path, SIZE_MAX, &buf, NULL);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to read sysctl value '%s': %m", path);
+
+        delete_trailing_chars(buf, "\n");
+
+        if (!streq(value, buf))
+                log_warning("sysctl %s was changed from '%s' to '%s'", path, value, buf);
+
+        return 0;
+}
+
+static int manager_inotify_add(Manager *manager) {
+        sd_event_source **source;
+        int r;
+
+        assert(manager);
+
+        source = manager->sysctl_event_sources;
+
+        FOREACH_STRING(p,
+                "/proc/sys/net/ipv4/conf/all",
+                "/proc/sys/net/ipv4/conf/default",
+                "/proc/sys/net/ipv6/conf/all",
+                "/proc/sys/net/ipv6/conf/default") {
+
+                r = sd_event_add_inotify(manager->event, source, p, IN_CLOSE_WRITE | IN_ONLYDIR, sysctl_inotify, manager->sysctl_shadow);
```

Comment:
i.e. for the pointer-to-pointer thing, just pass `&manager->sysctl_shadow` here as last arg